### DISCUSSION TITLE
How can I generate schema from a TypeScript type?

### DISCUSSION THREAD

**Discussion by babakfp** [open-circle/valibot#477] [Q&A] [answered]
Hi

```ts
import type { RehypeShikiOptions } from "@shikijs/rehype"
```

I want to use the above type in below schema:

```ts
        builtInPlugins: v.optional(
            v.object({
                shiki: v.optional(
                    v.object({
                        enable: v.optional(v.boolean(), true),
                        options: // Here!
                    }),
                    {}
                ),
            }),
            {}
        ),
```

**Comment by fabian-hiller** [comment]
As far as I know, there is currently no way to generate it dynamically, but you could use the following site to translate your type automatically to a Valibot schema: https://sinclairzx81.github.io/typebox-workbench/

**Reply by babakfp** [reply]
That's an awesome tool, however, the types are coming from other libraries.

Is it possible to at least get type auto-completion? This is what I'm currently doing:

```ts
export const ConfigSchema = v.optional(
    v.object({
        builtInPlugins: v.optional(
            v.object({
                remarkParse: v.optional(v.object({}), {}),
            }),
            {}
        ),
    }),
    {}
)

// I would like to remove this
export type ConfigInput = v.Input<typeof ConfigSchema> & {
    builtInPlugins?: {
        remarkParse?: {
            options?: RemarkParseOptions
        }
    }
}

// I would like to remove this
export type ConfigOutput = v.Output<typeof ConfigSchema> & {
    builtInPlugins: {
        remarkParse: {
            options?: RemarkParseOptions
        }
    }
}
```

**Reply by fabian-hiller** [reply]
Why don't you write it that way?

```ts
const ConfigSchema = v.optional(
  v.object({
    builtInPlugins: v.optional(
      v.object({
        remarkParse: v.optional(v.object({}), {}),
      }),
      {}
    ),
  }),
  {}
);

type ConfigInput = v.Input<typeof ConfigSchema>;
type ConfigOutput = v.Output<typeof ConfigSchema>;
```

**Reply by babakfp** [reply]
Because I need the type completions for `RemarkParseOptions`. Without the modifications in `ConfigInput` and `ConfigOutput` I won't be getting the auto-coalitions, JSDoc comments, and more.
In Zod, you can do `z.custom<MyType>(() => true, 'Error message')` and it does what I'm looking for.

**Reply by fabian-hiller** [reply, marked as answer]
We have the same function. Take a look: https://valibot.dev/api/custom/

**Reply by Kovinda** [reply]
This page seems to be missing from the docs...

**Reply by fabian-hiller** [reply]
Thank you for reaching out. I was renamed to [`custom`](https://valibot.dev/api/custom/) last year. I will update the comment.
